functions/src/index.ts

```diff
@@ -1,44 +1,53 @@
-import * as admin from "firebase-admin";
-import * as functions from "firebase-functions";
+import admin from "firebase-admin";
+import functions, { Change, EventContext } from "firebase-functions";
+import { DocumentSnapshot } from "firebase-functions/v1/firestore";
 import { v4 as uuidv4 } from "uuid";
 import { useAuthentication } from "./authentication";
 import { useCors } from "./cors";
 import { studies } from "./studies";
+import { isDeepStrictEqual } from "util";
+import * as gleanPings from "./glean";
 
 admin.initializeApp({
   credential: admin.credential.applicationDefault(),
 });
 
-export const rallytoken = functions.https.onRequest(async (request, response) =>
-  useCors(request, response, async () => {
-    await useAuthentication(request, response, async (decodedToken) => {
-      if (request.method !== "POST") {
-        response.status(500).send("Only POST and OPTIONS methods are allowed.");
-        return;
-      }
-
-      functions.logger.info(`body type: ${typeof request.body}`, {
-        payload: request.body,
-      });
-
-      try {
-        let studyId;
-        if (typeof request.body === "string") {
-          const body = JSON.parse(request.body);
-          studyId = body.studyId;
-        } else {
-          studyId = request.body.studyId;
-        }
-
-        const rallyToken = await generateToken(decodedToken, studyId);
-        functions.logger.info("OK");
-        response.status(200).send({ rallyToken });
-      } catch (ex) {
-        functions.logger.error(ex);
-        response.status(500).send();
-      }
-    });
-  })
+export const rallytoken = functions.https.onRequest(
+  async (request, response) =>
+    new Promise((resolve) =>
```

Comment:
Hm, why does this need the additional promise wrapper?

Reply:
This one is a little bit funny. The `cors` object is an old JavaScript function from Express that only uses a callback, not a promise. We need to be able to `await` properly on the `rallytoken` function for our tests. So I wrapped it in a promise and resolve it at the end of the callback.